crates/sc-proof-of-time/src/source/gossip.rs

```diff
@@ -1,26 +1,70 @@
 //! PoT gossip functionality.
 
-use crate::source::state::PotState;
+use crate::source::state::{NextSlotInput, PotState};
 use crate::verifier::PotVerifier;
 use futures::channel::mpsc;
+use futures::executor::block_on;
 use futures::{FutureExt, SinkExt, StreamExt};
+use lru::LruCache;
 use parity_scale_codec::{Decode, Encode};
 use parking_lot::Mutex;
 use sc_network::config::NonDefaultSetConfig;
-use sc_network::PeerId;
+use sc_network::{NetworkPeers, PeerId};
 use sc_network_gossip::{
     GossipEngine, MessageIntent, Network as GossipNetwork, Syncing as GossipSyncing,
     ValidationResult, Validator, ValidatorContext,
 };
+use sp_consensus::SyncOracle;
 use sp_consensus_slots::Slot;
 use sp_runtime::traits::{Block as BlockT, Hash as HashT, Header as HeaderT};
+use std::cmp;
+use std::collections::HashMap;
 use std::future::poll_fn;
-use std::num::NonZeroU32;
-use std::sync::atomic::Ordering;
-use std::sync::Arc;
-use subspace_core_primitives::{PotCheckpoints, PotSeed};
+use std::hash::{Hash, Hasher};
+use std::num::{NonZeroU32, NonZeroUsize};
+use std::sync::{atomic, Arc};
+use subspace_core_primitives::{PotCheckpoints, PotSeed, SlotNumber};
 use tracing::{debug, error, trace, warn};
 
+/// How many slots can proof be before it is too far
+const MAX_SLOTS_IN_THE_FUTURE: u64 = 10;
+/// How much faster PoT verification is expected to be comparing to PoT proving
+const EXPECTED_POT_VERIFICATION_SPEEDUP: usize = 7;
+const GOSSIP_CACHE_SIZE: NonZeroUsize = NonZeroUsize::new(1_000).expect("Not zero; qed");
+
+mod rep {
+    use sc_network::ReputationChange;
+
+    /// Reputation change when a peer sends us a gossip message that can't be decoded.
+    pub(super) const GOSSIP_NOT_DECODABLE: ReputationChange =
+        ReputationChange::new(-(1 << 3), "PoT: not decodable");
+    /// Reputation change when a peer sends us checkpoints that do not match next slot inputs.
+    pub(super) const GOSSIP_NEXT_SLOT_MISMATCH: ReputationChange =
```

Comment:
Does all these needs to have the same `ReputationChange`? 
For example `GOSSIP_SLOT_ITERATIONS_OUTSIDE_OF_RANGE` should deserve much lower reputation no ?

Reply:
The cost for all of these is an educated guess, but generally corresponds to a place of the corresponding check resulting in punishment in the pipeline. For all quick checks that happen earlier the pipeline the amount of computation done is much lower than for those check that happen much later and especially for those that require actual PoT verification.

Specifically `GOSSIP_SLOT_ITERATIONS_OUTSIDE_OF_RANGE` is subjective, it doesn't mean number of iterations isn't current, nodes can simply be out of sync, but one way or the other it doesn't match expectations and we have expended some resources to come up with that conclusion, hence small punishment.